File: 3171-find-subarray-with-bitwise-and-closest-to-k/3171-find-subarray-with-bitwise-and-closest-to-k.cpp
class Solution {
public:
    void changeBitCounts(int n, int flag, vector<int> &countBits) {
        int i = 0;
        while (n>0) {
            if (n&1) {
                countBits[i] += flag;
            }
            n = n/2;
            i++;
        }
    }
    
    
    int minimumDifference(vector<int>& nums, int k) {
        vector <int> countBits(32, 0);
        int ans = INT_MAX;
        int st = 0;
        int cur = nums[0];
        
        for(int i=0; i<nums.size(); i++) {
            cur = cur & nums[i];
            changeBitCounts(nums[i], 1, countBits);
            ans = min(ans, abs(k-cur));
            
            if (k == cur) return 0;
            
            if (cur>k) continue;
            
            else {
                while (st<=i && cur<k) {
                    changeBitCounts(nums[st], -1, countBits);
                    st++;
                    
                    cur = 0;
                    for (int j=0; j<32; j++) {
                        if (i-st+1 == countBits[j]) {
                            cur = cur | (1<<j);
                        }
                    }
                ans = min(ans, abs(k-cur));
                }
                
            }
        }
        
    return ans;
    }
};

File: 3171-find-subarray-with-bitwise-and-closest-to-k/3171-find-subarray-with-bitwise-and-closest-to-k.py
class Solution(object):
    def changeBitCounts(self, n, flag, countBits):
        i = 0
        while n > 0:
            if n & 1:
                countBits[i] += flag
            n = n // 2
            i += 1
    
    
    def minimumDifference(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        countBits = [0 for _ in range(32)]
        ans = sys.maxint
        st = 0
        cur = nums[0]
        
        for i in range(len(nums)):
            cur = cur & nums[i]
            self.changeBitCounts(nums[i], 1, countBits)
            ans = min(ans, abs(k-cur))
            
            if cur == k:
                return 0
            
            elif cur > k:
            # As numbers needs to be added to the subarray to decrease the value of cur to bring it nearer to k
                continue
            else:
                while st<=i and cur < k:
                    # Removing element by element till i
                    self.changeBitCounts(nums[st], -1, countBits)
                    st += 1
                    
                    cur = 0
                    for j in range(32):
                        # i-st+1 is nothing but the remaining size of the sub array being considered. so if it equals to countBits[j] that mean that bit is one for all the elements of that subarray so it can be considered one in the cur and others as zero.
                        if (i-st+1) == countBits[j]:
                            cur = cur | (1<<j)
                            
                    ans = min(ans, abs(k-cur))
        
        return ans